pandas/tests/test_strings.py

```diff
@@ -34,11 +188,115 @@ def test_api(self):
         assert Series.str is strings.StringMethods
         assert isinstance(Series(['']).str, strings.StringMethods)
 
-        # GH 9184
-        invalid = Series([1])
-        with pytest.raises(AttributeError, match="only use .str accessor"):
-            invalid.str
-        assert not hasattr(invalid, 'str')
+    @pytest.mark.parametrize('dtype', [object, 'category'])
+    @pytest.mark.parametrize('box', [Series, Index])
+    def test_api_per_dtype(self, box, dtype, all_skipna_inferred_dtypes):
+        # one instance of parametrized fixture
+        inferred_dtype, values = all_skipna_inferred_dtypes
+
+        t = box(values, dtype=dtype)  # explicit dtype to avoid casting
```

Comment:
you can actually make this 2 tests by putting the xfails in a function. then the test becomes single purpose and you don't have the if statement near the bottom.

Reply:
I don't understand what you're asking for here, sorry.

The test is already very single-purpose (except the xfails, which will be gone with #23167 and follow-up PRs), and the final if-switch makes it transparent which types are actually passing the constructor, with all other types raising. this would only get harder to understand if it's split into two, no?

Reply:
no you have an if else branch. it makes reasoning about this impossible as its very fixture / data dependent.

Reply:
The if branch is transparently for the cases where the `.str`-accessor raises or not. I do not understand how you want me to structure this test (resp. this function you mentioned).

Reply:
if/else make the test very hard to understand. pls break in 2

Reply:
This particular test can be broken into `_passes` and `_raises`, but that last if condition is really not that hard. Don't get that objection, tbh.
```
if inferred_dtype in types_passing_constructor:
    # pass
else:
    # raise

Reply:
Respectfully, I find it an unrealistic criterion to not be able to use *one* simple if-condition in a test (aside from xfails, which will be gone soon after). The whole point is that it's got an extensively parametrized fixture (`any_skipna_inferred_dtype`), and I have to make a single distinction based on the content of that fixture.

Even if I were to split this test into `_passes` and `_raises`, I'd have to use the *same* kind of if-condition, unless I were to needlessly break up the parametrized fixtures into smaller subsets.